src/service/resources-impl.js

```diff
@@ -239,43 +278,147 @@ export class ResourcesImpl {
+          return;
+        }
+
+        if (r.hasOwner()) {
+          return;
+        }
+
+        // For just-unloaded resources, setInViewport() will be called
+        // as part of Resource.unlayout().
+        r.setInViewport(isIntersecting);
+
+        // TODO(willchou): The lack of "update on scroll throttling" means
+        // that scrolled-over elements are no longer deferred, which results
+        // in longer delays for in-viewport elements after fast scrolling.
+        // Fix by queueing intersection entries when scroll velocity is high.
+        if (isIntersecting && r.isDisplayed()) {
+          if (r.getState() === ResourceState.READY_FOR_LAYOUT) {
+            this.scheduleLayoutOrPreload(r, /* layout */ true);
+          }
+        }
+      });
+    });
+
+    Promise.all(promises).then(() => {
+      this.unloadResources_(toUnload);
+      this.signalIfReady_();
+    });
+  }
+
   /** @private */
   rebuildDomWhenReady_() {
     // Ensure that we attempt to rebuild things when DOM is ready.
     this.ampdoc.whenReady().then(() => {
       this.documentReady_ = true;
       this.buildReadyResources_();
       this.pendingBuildResources_ = null;
-      const fixPromise = ieMediaCheckAndFix(this.win);
-      const remeasure = () => this.remeasurePass_.schedule();
-      if (fixPromise) {
-        fixPromise.then(remeasure);
-      } else {
-        // No promise means that there's no problem.
-        remeasure();
-      }
+
       const input = Services.inputFor(this.win);
       input.setupInputModeClasses(this.ampdoc);
 
-      // Safari 10 and under incorrectly estimates font spacing for
-      // `@font-face` fonts. This leads to wild measurement errors. The best
-      // course of action is to remeasure everything on window.onload or font
-      // timeout (3s), whichever is earlier. This has to be done on the global
-      // window because this is where the fonts are always added.
-      // Unfortunately, `document.fonts.ready` cannot be used here due to
-      // https://bugs.webkit.org/show_bug.cgi?id=174030.
-      // See https://bugs.webkit.org/show_bug.cgi?id=174031 for more details.
-      Promise.race([
-        loadPromise(this.win),
-        Services.timerFor(this.win).promise(3100),
-      ]).then(remeasure);
-
-      // Remeasure the document when all fonts loaded.
-      if (
-        this.win.document.fonts &&
-        this.win.document.fonts.status != 'loaded'
-      ) {
-        this.win.document.fonts.ready.then(remeasure);
+      // With IntersectionObserver, no need for remeasuring hacks.
+      if (!this.intersectionObserver_) {
+        const fixPromise = ieMediaCheckAndFix(this.win);
+        const remeasure = () => this.remeasurePass_.schedule();
+        if (fixPromise) {
+          fixPromise.then(remeasure);
+        } else {
+          // No promise means that there's no problem.
+          remeasure();
+        }
+
+        // Safari 10 and under incorrectly estimates font spacing for
```

Comment:
Probably for a separate PR, but we no longer support Safari 10, so if this is no longer an issue, we should clean up.